Array cell sensitivity

romainbrenguier opened the pull request:
This pull request activates the array field sensitivity code, which should allow constant propagation in arrays of constant size.
 
<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

smowton committed:
Add L1_WITH_CONSTANT_PROPAGATION mode to goto_symex_statet::rename

This allows applying constant propagation without upgrading non-constant symbols to L2,
which is useful for the dereferencing logic, as the value-set used to resolve pointer
aliasing is indexed by L1 names.

smowton committed:
Symex: enable array cell sensitivity

smowton committed:
Array cell sensitivity: tolerate zero-length arrays

smowton committed:
Resolve array indices before dereferencing

This may lead to application of array-cell-sensitivity (i.e. querying the symbol
some_array[[1]] instead of some_array[some_index]), leading to higher precision.

Some tests must be amended because better constant propagation causes some test behaviour
to change as symex has better knowledge regarding the targets of pointers.

romainbrenguier committed:
Set a limit size for array cell propagation

The use of field_sensitivity for arrays can be expensive for big arrays
so we have to set a limit on the size of constant arrays which get
processed.
The value 64 was determined by changing the size of the array in the
test regression/cbmc/variable-access-to-constant-array and reducing it
until there was no significant difference in execution time between the
versions of CBMC with and without array cell propagation.
For the chosen value the execution time was around 0.05 second in both
cases (for comparison with size 1024 the time with propagation was 0.5
sec, against 0.1 without).

romainbrenguier committed:
Attempt to get missing array size from symbol table

In some cases, like the test in regression/cbmc/Global_Initialization2,
the array type is incomplete and changed in the symbol table after the
main function has been converted, leading to inconsistencies.
This means we can get nil instead of the size in
field_sensitivityt::apply, though the actual size is present in the
symbol table.
The issue is reported here: https://github.com/diffblue/cbmc/issues/5022

romainbrenguier committed:
Rename size to L2 in field_sensitivity::apply

This is to allow constant propagation to take place and know whether the
size is actually constant, which can allow field sensitivity to apply.

romainbrenguier committed:
Make field_sensitivity::apply handle non-constant index

When the index is not constant when accessing a constant size array, we
should expand the expression into `{array[0]; array[1];...}[index]` so
that the relation with field sensitivity expressions `array[[0]]`,
`array[[1]]`, etc, is not lost.

romainbrenguier committed:
Only apply field_sensitivity to dereferenced object

This is important to consider dereferenced object as read value.

romainbrenguier committed:
Apply field sensitivity in symex_printf

This may be necessary for constant propagation of the format string.
An example of that is in regression/cbmc/printf1

romainbrenguier committed:
Allow trace to assign in two steps

With field sensitivity on arrays, the element can be initialized in two
steps, first the field f then the field array.

romainbrenguier committed:
Adapt double_deref test for array cell sensitivity

Propagatiof of values of array cells lead to different expressions in VCC.

romainbrenguier committed:
Tag test failing with SMT backend

This is failing because of this bug:
https://github.com/diffblue/cbmc/issues/4749

romainbrenguier committed:
Simplify lhs in build goto trace

Avoid lhs of the form `{array[0], array[1]}[0]`

smowton committed:
Add tests for array-cell sensitivity

Most are for C, then a couple are duplicated for Java to check whether its use of a structure
to represent all arrays makes any difference.

romainbrenguier committed:
Document how arrays are handled by field sensitivity

This explains how field sensitivity transforms instructions that contain
array operations.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 4b5023f9716980216f570827977e7047fa944261).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/122071988

romainbrenguier commented:
I ran some experiments on 250 methods from tika-parsers (with unwind 15). It seems it doesn't help on simple cases (even a bit slower) but on some more complicated examples it is much faster.
I tested this with two version of models for the standard library, one uses non-determinstic size arrays and the other one uses constant size arrays so that we can make more use of array cell sensitivity.
   
![Develop, Models and Array cell sensitivity](https://user-images.githubusercontent.com/11854194/62530144-e79c4100-b837-11e9-8e2d-10ae96a63c59.png)
(time in sec.)

https://docs.google.com/spreadsheets/d/e/2PACX-1vRJNu9_cmWBc2ptmMLNWkU2895p-Kk40XGPnGU6yuP4eI4okC5muU0S9dm-u7Wxyc5mqQlrraW3qogD/pubchart?oid=830404797&format=interactive

tautschnig commented on src/goto-symex/field_sensitivity.cpp:
```diff
@@ -97,10 +97,21 @@ exprt field_sensitivityt::apply(
+      {
+        // In case the array type was incomplete, attempt to retrieve it from
+        // the symbol table.
+        const symbolt *array_from_symbol_table = ns.get_symbol_table().lookup(
```
What is the justification for this to be correct? I don't see why we could use values from the symbol table during symbolic execution.

romainbrenguier replied:
I've only seen that happen in one case which is this regression test: https://github.com/diffblue/cbmc/blob/develop/regression/cbmc/Global_Initialization2/main.c
In this example since `A1` is declared but not defined the length we get at this point in symex is not defined, although it appears correctly in the symbol table. @peterschrammel seemed to think that could be caused by a bug in the C frontend but we couldn't identify the source of this issue.

peterschrammel replied:
As far as I remember the problem is that the array type is created with undefined length and then attached to various expressions from which the goto program is constructed. When The implementation of rule 6.9.2 (https://github.com/diffblue/cbmc/blob/develop/regression/cbmc/Global_Initialization2/main.c#L15) then updates the array length of the symbol in the symbol table only, but does not end up in the goto program. Symex then takes the length from the goto program, which doesn't contain the correct length - the correct length is in the symbol table only.
IMO there are two solutions for that: 1. get rid of the length in the array type. 2. In the C frontend, process declarations first to see whether all array types are complete, then update; only afterwards convert function bodies.

peterschrammel replied:
@romainbrenguier, please create a ticket in the issue tracker for this so that we can go ahead.

romainbrenguier replied:
issue created here https://github.com/diffblue/cbmc/issues/5022, I will add it to the commit message

tautschnig reviewed (commented):
Just trying to make this manageable:
1) One comment on something that surprised me below.
2) Could any commits be factored out and be merged independently?
3) What is the timeline along which we are hoping to merge this?

tautschnig commented on regression/cbmc/address_space_size_limit3/main.i:
```diff
@@ -92,6 +92,11 @@ void f__L_0x3c6_0()
+  // Avoid constant propagation solving for pointer-offsets of esp,
+  // which prevent demonstrating the object-size limitation this test
+  // intends to exhibit:
+  esp -= (unknown & 1);
```
This actually changes the semantics of the program? Maybe do `esp += (unknown & 1);` to counter that?

tautschnig commented on regression/cbmc/array-cell-sensitivity9/test.desc:
```diff
@@ -0,0 +1,32 @@
+CORE
+test.c
+--show-vcc
```
It seems that all the tests only exercise `--show-vcc` - I think we should also be running the full flow (in addition to `--show-vcc`) to make sure we actually get the right results!

tautschnig commented on src/goto-symex/field_sensitivity.h:
```diff
@@ -9,6 +9,10 @@ Author: Michael Tautschnig
 
+/// Limit the size of arrays for which field_sensitivity gets applied.
+/// Necessary because large constant arrays slow-down the process.
+static constexpr unsigned max_field_sensitive_array_size = 64;
```
This should go in `util/magic.h`

tautschnig commented:
There are some `KNOWNBUG` tests in `regression/cbmc-concurrency` - I believe at least `struct_and_array1` should be fixed by this PR. The others are also worth exploring.

tautschnig commented:
@romainbrenguier @smowton Thanks you for building this! I've done a high-level review with a bunch of questions.

romainbrenguier committed:
Activate regression tests previously KNOWNBUG

These tests can now be executed in our test suite.

romainbrenguier commented:
@tautschnig 
> There are some KNOWNBUG tests in regression/cbmc-concurrency - I believe at least struct_and_array1 should be fixed by this PR. The others are also worth exploring.

The test doesn't work on windows:
`main.c(1): fatal error C1083: Cannot open include file: 'pthread.h': No such file or directory `
Do you have an idea of how to fix that?

smowton commented:
Tag the test `gcc-only`, and check the Makefile for that directory is generating an exclusion (`-X gcc-only`) when running under MSVC. See regression/cbmc/Makefile for an example.

smowton commented:
Benchmarking with SVCOMP shows no particular performance difference, but does uncover a probable bug: test `linux-4.0-rc1---drivers--net--ethernet--ethoc.ko.cil.i` exits with `l2_rename_rvalues case 'constant' not handled`

smowton commented:
Minimised bug to

```
int main(int argc, char **argv) {

  int x;
  const char *c = "Hello world";

  int *p = (argc ? &x : (int *)c);

  *p = 1;

  return 0;

}
```

smowton commented:
That example faults on current develop; fix at https://github.com/diffblue/cbmc/pull/4998

smowton committed:
Add execution tests for array-cell-sensitivity

These are companions to the existing tests that check the show-vcc output, and serve to check
that the formulas generated appear to behave properly.

peterschrammel commented on src/util/magic.h:
```diff
@@ -16,4 +16,8 @@ const std::size_t MAX_CONCRETE_STRING_SIZE = 1 << 26;
 
+/// Limit the size of arrays for which field_sensitivity gets applied.
+/// Necessary because large constant arrays slow-down the process.
+constexpr unsigned max_field_sensitive_array_size = 64;
```
Any reason why this is `unsigned` and not `size_t`?

peterschrammel replied:
Any reason why this is lower case and not all-caps as all the other constants defined in this file?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a8a3e941df9a891ebbda25786446b2c8d30305d7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123201104

peterschrammel reviewed (approved):
modulo fixing clang and doxygen issues

romainbrenguier commented:
> Just trying to make this manageable:
> 1) One comment on something that surprised me below.

I've now raised an issue for it 

> 2) Could any commits be factored out and be merged independently?

I've already done that for as many things as I could, I don't think there is much we can extract now that would make sense on its own

> 3) What is the timeline along which we are hoping to merge this?

We think it's ready now, so as soon as possible.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 53e8c93c78ed97cdac3223bff787596273e360bc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/123473784

romainbrenguier committed:
Make field sensitivity max array size configurable

This will allow the user of symex to set this limit as needed instead of
fixing it to an arbitrary value.

romainbrenguier committed:
Add max-field-sensitivity-array-size option

together with no-array-field-sensitivity, this makes it possible to
control the maximum size for which the transformation is applied and
totally deactivate field sensitivity for arrays if needed.

romainbrenguier committed:
Add tests for field sensitivity options

This checks that field sensitivity applies or not to arrays depending on
their size and the value of the option.

romainbrenguier commented:
@peterschrammel @smowton I've added options to choose the maximum length for which field sensitivity applies (two last commits)

smowton commented on src/goto-symex/symex_main.cpp:
```diff
@@ -46,7 +46,11 @@ symex_configt::symex_configt(const optionst &options)
     show_symex_steps(options.get_bool_option("show-goto-symex-steps")),
     max_field_sensitivity_array_size(
-      options.get_unsigned_int_option("max-field-sensitivity-array-size"))
+      options.is_set("no-array-field-sensitivity")
```
We should produce an error if both are set

romainbrenguier replied:
I've added a check for that in jbmc/cbmc_parse_options

smowton commented:
Also genuine CI problems to fix; after that let's merge!

smowton commented:
Merged as #5057
